amp-pwa-reader/gulpfile.js

```diff
@@ -131,15 +131,69 @@ function modularizeJS(name) {
   return "\nmodule.exports = " + name + ';';
 }
 
-var dist = gulp.series(gulp.parallel(copy, styles, scripts, server), inline, injectManifest);
+/* Generates a complete ServiceWorker, with precaching and runtime caching rules. */
+function buildSW() {
+    // This will return a Promise
+    return workboxBuild.generateSW({
+        globDirectory: './dist/',
+        // Static precaching of shell
+        globPatterns: [
+            'img/*.{svg,png,jpg}', 'index.html', 'inline.css'
+        ],
+        swDest: './dist/sw.js',
+        // Register main route for all navigation links to pages
+        navigateFallback: 'index.html',
+        navigateFallbackBlacklist: [/img\/.*/, /\.(js|css)/],
+        navigateFallbackWhitelist: [/./],
+        // Cache external libraries and fonts
+        runtimeCaching: [{
+                urlPattern: new RegExp('^https://cdn\.ampproject\.org/'),
+                handler: 'staleWhileRevalidate',
+            },
+            {
+                urlPattern: new RegExp('^https://cdn\.polyfill\.io/'),
+                handler: 'staleWhileRevalidate',
+            },
+            {
+                urlPattern: new RegExp('^https://pasteup\.guim\.co\.uk/fonts/'),
+                handler: 'cacheFirst',
+            },
+            // Cache a number of YQL queries, but only for the offline scenario
+            {
+                urlPattern: new RegExp('^https://query\.yahooapis\.com/v1/public/'),
+                handler: 'networkFirst',
+            },
+            // Cache a number of images
+            {
+                urlPattern: new RegExp('^https://i\.guim\.co\.uk/img/'),
+                handler: 'cacheFirst',
+                options: {
+                    cacheName: 'images',
+                    expiration: {
+                        maxEntries: 10,
+                        maxAgeSeconds: 7 * 24 * 60 * 60,
+                    },
+                    cacheableResponse: {
+                        statuses: [0, 200],
+                    },
+                },
+            },
+        ],
+        // Make sure new versions of the Service Worker activate immediately
+        clientsClaim: true,
+        skipWaiting: true,
+    });
```

Comment:
Can you add in something that will log the precache size any warnings? This was left out of the official docs; I've filed https://github.com/google/WebFundamentals/pull/6158 to update those, and you could pretty much add the same thing here.